common/lib/client/channel.ts

```diff
@@ -0,0 +1,178 @@
+}
+
+function noop() {}
+const MSG_ID_ENTROPY_BYTES = 9;
+
+function allEmptyIds(messages: Array<Message>) {
+	return Utils.arrEvery(messages, function(message: Message) {
+		return !message.id;
+	});
+}
+
+function normaliseChannelOptions(options?: ChannelOptions) {
+	const channelOptions = options || {};
+	if(channelOptions.cipher) {
+		if(!Crypto) throw new Error('Encryption not enabled; use ably.encryption.js instead');
+		const cipher = Crypto.getCipher(channelOptions.cipher);
+		channelOptions.cipher = cipher.cipherParams;
+		channelOptions.channelCipher = cipher.cipher;
+	} else if('cipher' in channelOptions) {
+		/* Don't deactivate an existing cipher unless options
+			* has a 'cipher' key that's falsey */
+		channelOptions.cipher = null;
+		channelOptions.channelCipher = null;
+	}
+	return channelOptions;
+}
+
+class Channel extends EventEmitter {
+	rest: Rest | Realtime;
+	name: string;
+	basePath: string;
+	presence: Presence;
+	channelOptions: ChannelOptions;
+
+	constructor(rest: Rest | Realtime, name: string, channelOptions?: ChannelOptions) {
+		super();
+		Logger.logAction(Logger.LOG_MINOR, 'Channel()', 'started; name = ' + name);
+		this.rest = rest;
+		this.name = name;
+		this.basePath = '/channels/' + encodeURIComponent(name);
+		this.presence = new Presence(this);
+		this.channelOptions = {};
+		this.channelOptions = normaliseChannelOptions(channelOptions);
+	}
+
+	setOptions(options: ChannelOptions): void {
+		this.channelOptions = normaliseChannelOptions(options);
+	}
+
+	history(params: RestHistoryParams | null, callback: PaginatedResultCallback<Message>): Promise<PaginatedResult<Message>> | void {
+		Logger.logAction(Logger.LOG_MICRO, 'Channel.history()', 'channel = ' + this.name);
+		/* params and callback are optional; see if params contains the callback */
+		if(callback === undefined) {
+			if(typeof(params) == 'function') {
+				callback = params;
+				params = null;
+			} else {
+				if(this.rest.options.promises) {
+					return Utils.promisify(this, 'history', arguments);
+				}
+				callback = noop;
+			}
+		}
+
+		this._history(params, callback);
+	}
+
+	_history(params: RestHistoryParams | null, callback: PaginatedResultCallback<Message>): void {
+		const rest = this.rest,
+			format = rest.options.useBinaryProtocol ? Utils.Format.msgpack : Utils.Format.json,
+			envelope = Http.supportsLinkHeaders ? undefined : format,
+			headers = Utils.defaultGetHeaders(format);
+
+		if(rest.options.headers)
+			Utils.mixin(headers, rest.options.headers);
+
+		const options = this.channelOptions;
+		(new PaginatedResource(rest, this.basePath + '/messages', headers, envelope, function(body: any, headers: Record<string, string>, unpacked?: boolean) {
+			return Message.fromResponseBody(body, options, unpacked ? undefined : format);
+		})).get(params as Record<string, unknown>, callback);
+	}
+
+	publish(): void | Promise<void> {
+		const argCount = arguments.length,
+			first = arguments[0],
+			second = arguments[1];
```

Comment:
Is it safe to directly access arguments[N]?

Reply:
Yup, if you don't provide an argument here then it will return `undefined` and we handle all those cases below

Reply:
Resolving this conversation on @ikbalkaya's behalf.